### DISCUSSION TITLE
Report - (3rd Sem.) Week 6 📢

### DISCUSSION THREAD

**Discussion by maxschwinghammer** [SE-TINF22B6/DHBWhub#165] [Weekly Report]
Welcome to the sixth report of our progress. We are excited to share the latest news with you! 🥳

#### 🚀 Project Updates:

We have successfully implemented the Post component in React, enriching our frontend capabilities.
![image](https://github.com/SE-TINF22B6/DHBWhub/assets/111356450/439324b8-8aab-4fb3-98fa-6ec9763f1ba7)

#### We also created a mockup for the contact page in Figma.
🖥️PC view:
![Contact - PC View](https://github.com/SE-TINF22B6/DHBWhub/assets/111356450/5feacd30-252a-43d3-a211-085175a0d6cf)
📱Mobile view:
![Contact - Smartphone View](https://github.com/SE-TINF22B6/DHBWhub/assets/111356450/fef09fdb-3392-429b-815c-a97672e6904a)

#### 🛠️ System Architecture Development:
In this week's developments, our system architecture has taken shape through an Object Modeling Language (OML) diagram, embodying the principles of the **Model-View-Controller** (MVC) design pattern with a nod to **Interface Segregation**. By adopting the MVC design pattern, we aim to achieve a scalable, maintainable, and modular system architecture. This separation of concerns allows for a more organized development process, making it easier to collaborate, maintain, and extend our project in the future.   
In the following diagram, you can see the whole structure which might seem confusing at start, but we will take a closer look at one example:
   
![uml_class_diagram_improved drawio (1)](https://github.com/SE-TINF22B6/DHBWhub/assets/111356450/0c2de9b3-d3b1-4b01-807f-5612b1d2d953)
![example](https://github.com/SE-TINF22B6/DHBWhub/assets/122597204/5ac61594-1b87-42e3-b106-57a12748a90d)

**Key Entities:**

- **Account Entity:** Representing user accounts, the Account entity encompasses attributes like account_id, username, email, and password. We've implemented an AccountService interface, with its practical application employing SpringAccountRepository.
- **Client User Entity:** This entity mirrors user accounts but includes additional features such as age and description. Controllers and services for seamless management have been integrated.
- **Administrator Entity:** Specifically tailored for administrators, this entity is equipped with attributes like admin_id. Operations are streamlined through the AdministratorService interface.
- **Picture Entity:** Capturing images with attributes like picture_id, name, and pictureData, the Picture entity is managed through the defined operations in the PictureService.
- **Post Entity:** Our system handles posts elegantly with attributes like post_id, title, description, date, likes, and tags. Operations are meticulously outlined in the PostService interface.
- **Comment Entity:** Comments are seamlessly incorporated with attributes like comment_id, description, and date. Controllers and services are set up for efficient comment management.
- **Course and Faculty Entities:** For courses and faculties, we've established controllers and services, adhering to standard CRUD operations.

**Entity controllers:** Each entity has a controller, which is the endpoint the web application uses. In order to retrieve the requested data from the database, the controller uses a service interface.
  
**Service Interfaces and Implementations:** Each entity has its dedicated service interface (e.g., AccountService, ClientUserService) and a corresponding implementation (e.g., AccountServiceImplementation, ClientUserServiceImplementation). The service is responsible for the 
business logic and uses a repository class for database retrievals. A service class of one entity may also use the service class of another entity as you can see for the entities "client user" and "administrator" which both use the account service.

**Repositories:** The repository for each entity is responsible for basic CRUD-operations but also specific SQL-commands on the database. For best practice, we use wrapper classes which use the raw repository entity to ensure more safety and segregation between the services and the repositories.

### Sprint 3: What have we achieved ❓ What's to come❔
✅ Creation of the UML diagram
✅ First draft of the calendar
✅ More Spring Boot tests
✅ Implementation of Post View
✅ Mockup for contact page
🚧 Finish Calendar Mockup in Figma
🚧 Implement Login Window
🚧 Implement Sign Up Window
🚧 Discuss search opportunities and create mockup
🚧 Implement post menu
🚧 Implement contact page
🚧 Implement first frontend endpoints
🚧 Implement e-mail verification
🚧 Setup postman for whole team
🚧 Create endpoints for search
🚧 Create endpoints for the sort options of posts
🚧 Removing Herobrine

Until next week 💯

Your DHBWhub Team 🚀

**Comment by LenVej** [comment]
Hey DHBWhub Team,

First off all I really like your design! It's simple and fits the kind of website you're going for. It is also great to see that you already implemented the Post component.

Furthermore, your Object Modeling Language (OML) diagram effectively captures the essence of the Model-View-Controller (MVC) design pattern, showcasing a thoughtful nod to Interface Segregation for a well-rounded architecture. The dedication to achieving a scalable, maintainable, and modular system is evident in your implementation of the MVC design pattern, fostering an organized development process.

Your CookHub Team.

**Comment by LaggyNacho** [comment]
The DHBWhub Team's latest update is cool! Shwoing the MVC design pattern did add improvement and brings more clarity. Also, I ike the Post component lot. Same goes to the UML diagram - it is pretty detailed. Also the idea to design the mobile view and Desktop view together may make things while designing easier, which is a good point to note aswell. Nice to see such progress with technical and design aspects.
